fw/preamp/src/front_panel.c

```diff
@@ -25,6 +25,7 @@
 
 bool audio_power_on = false;
 
+// Simple counter used for time-sensitive operations
 void delay(int a) {
```

Comment:
In general this is not a good idea if a consistent delay is needed. The time needed to execute this can vary based on flash read times, pipelining, and probably other things I don't know about in this architecture. This is simple enough it's _probably_ fairly consistent, but I'd have to look at the generated assembly and then see what the M0 architecture is doing. Adding instruction barriers might make things more consistent too. BUT, why dig into all of that when there's a timer designed for _exactly_ this function! Ok fine, it's designed for real-time systems like FreeRTOS but it's a very simple timer that'd be great for just "delay x us/ms/s/whatever"

Reply:
What is this magical timer?

Reply:
// Assumes the CPU's HCLK is 8 MHz from HSI,
// so this function is not robust AT ALL to changes in HCLK
// Max of 2^24, or ~16.7s with 8 MHz / 8
void delay_us(uint32_t us) {
  // Disable SysTick and its interrupts and set the clock source to HCLK / 8
  SysTick->CTRL &= ~(SysTick_CTRL_ENABLE_Msk | SysTick_CTRL_TICKINT_Msk | SysTick_CTRL_CLKSOURCE_Msk);
  SysTick->LOAD = nus;                        // Set the time to count down from
  SysTick->VAL = 0;                           // Clear the counter
  SysTick->CTRL |= SysTick_CTRL_ENABLE_Msk;   // Start countdown (and read CTRL which resets COUNTFLAG)

  // Keep reading CTRL until either SysTick is disabled or has counted to 0
  uint32_t ctrl;
  do {
    ctrl = SysTick->CTRL;
  } while((ctrl & SysTick_CTRL_ENABLE_Msk) && !(temp & SysTick_CTRL_COUNTFLAG_Msk));

  SysTick->CTRL &= ~SysTick_CTRL_ENABLE_Msk;  // Disable the counter
  SysTick->VAL = 0X00;                        // Clear the counter
}

References:
https://developer.arm.com/documentation/dui0497/a/cortex-m0-peripherals/optional-system-timer--systick/systick-control-and-status-register
https://www.st.com/en/microcontrollers-microprocessors/stm32f030r8.html#documentation (RM0360 and DS9773)